auth_plugin.go

```diff
@@ -0,0 +1,63 @@
+// Go MySQL Driver - A MySQL-Driver for Go's database/sql package
+//
+// Copyright 2023 The Go-MySQL-Driver Authors. All rights reserved.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package mysql
+
+// AuthPlugin represents an authentication plugin for MySQL/MariaDB
+type AuthPlugin interface {
+	// GetPluginName returns the name of the authentication plugin
+	GetPluginName() string
+
+	// InitAuth initializes the authentication process and returns the initial response
+	// authData is the challenge data from the server
+	// password is the password for authentication
+	InitAuth(authData []byte, cfg *Config) ([]byte, error)
+
+	// ContinuationAuth processes the authentication response from the server
+	// packet is the data from the server's auth response
+	// authData is the initial auth data from the server
+	// conn is the MySQL connection (for performing additional interactions if needed)
+	ContinuationAuth(packet []byte, authData []byte, conn *mysqlConn) ([]byte, error)
```

Comment:
Can external plugins use `mysqlConn`?

We should not expose plugin interface until its design is complete and stable.

Reply:
changed